server/admin_management/api/app/routers/router_application_admin.py

```diff
@@ -0,0 +1,168 @@
+    request: Request,
+    db: Session = Depends(database.get_db),
+    token_claims: dict = Depends(jwt_validation.authorize),
+    requester: Requester = Depends(get_current_requester),
+):
+
+    LOGGER.debug(
+        f"Executing 'create_user_role_assignment' "
+        f"with request: {application_admin_request}, requestor: {token_claims}"
+    )
+
+    audit_event_log = AuditEventLog(
+        request=request,
+        event_type=AuditEventType.CREATE_APPLICATION_ADMIN_ACCESS,
+        event_outcome=AuditEventOutcome.SUCCESS,
+    )
+
+    try:
+        application_admin_service = ApplicationAdminService(db)
+        application_service = ApplicationService(db)
+        user_service = UserService(db)
+
+        audit_event_log.requesting_user: models.FamUser = (
+            user_service.get_user_by_cognito_user_id(requester.cognito_user_id)
+        )
+        audit_event_log.application: models.FamApplication = (
+            application_service.get_application(
+                application_admin_request.application_id
+            )
+        )
+        audit_event_log.target_user: models.FamUser = (
+            user_service.get_user_by_domain_and_name(
+                application_admin_request.user_type_code,
+                application_admin_request.user_name,
+            )
+        )
+
+        return application_admin_service.create_application_admin(
+            application_admin_request, requester.cognito_user_id
+        )
+
+    except Exception as e:
+        audit_event_log.event_outcome = AuditEventOutcome.FAIL
+        audit_event_log.exception = e
+        raise e
+
+    finally:
+        if audit_event_log.target_user is None:
+            audit_event_log.target_user = models.FamUser(
+                user_type_code=application_admin_request.user_type_code,
+                user_name=application_admin_request.user_name,
+                user_guid="unknown",
+                cognito_user_id="unknown",
+            )
+
+        audit_event_log.log_event()
+
+
+@router.delete(
+    "/{application_admin_id}",
+    response_class=Response,
+    dependencies=[Depends(authorize_by_fam_admin), Depends(enforce_self_grant_guard)],
+)
+def delete_application_admin(
+    application_admin_id: int,
+    request: Request,
+    db: Session = Depends(database.get_db),
+    requester: Requester = Depends(get_current_requester),
+):
+    LOGGER.debug(
+        f"Executing 'delete_application_admin' with request: {application_admin_id}"
+    )
+
+    audit_event_log = AuditEventLog(
+        request=request,
+        event_type=AuditEventType.REMOVE_APPLICATION_ADMIN_ACCESS,
+        event_outcome=AuditEventOutcome.SUCCESS,
+    )
+
+    try:
+        application_admin_service = ApplicationAdminService(db)
+        user_service = UserService(db)
+
+        application_admin = application_admin_service.get_application_admin_by_id(
+            application_admin_id
+        )
+        if application_admin:
```

Comment:
um.... yeah we can do it in the router guard, but I feel router_guards is more for security and permission check? In this case, it's just a regular error case? and we might still want the log? I'm not so sure which way is better

Reply:
FastAPI's intention for **"Depends"** does more than just auth/security as you can see from the doc, can also be the path/param validations:
![image](https://github.com/bcgov/nr-forests-access-management/assets/81595625/a868bbf9-9705-4c2f-8b25-785044815953)
Initially the dependencies are in all in one file (**jwt_validation.py**), and was refactored into "**router_guards**" and leaving specific jwt validation to only in **jwt_validation.py**. 
The "router_guards" name may be a little sounds like just for "security/permission" but its more for **router's needs** (dependencies), as you can see in router_guards, there is a common code for getting the "**requester**" which most endpoints depend on.
In this case, the **{path /application_admin_id}** is meant to be **required and exists** for this endpoint to make sense , at the end, you would throw 400 if this id isn't valid; making it early check before the endpoint pass to business logic make sense.

This is my view, normally, the router/controller should be suppose to just dispatches all request params/body information to the service to handle and it should only at the end do some very less or even no logic at all regarding to business operation and simply forwarding the response back. 

We already have very large code not necessary for the "**audit**" (it should be somehow somewhere taking care of all endpoints' audit and endpoint just need to pass information to it; but we don't have that yet). Adding so many ad-hoc code at the router level will eventually becoming duplication everywhere.

It could also be done in another way. If you don't feel like to put the logic in **router_guards**, then as mentioned few days ago for the "**layered pattern**", commonly it could be seen in **"service" layer** it has a "**validator**" **function or class** taking care of these validations before executing business transactions.

In this case, I do feel FastAPI's "**Depends**" is a (and probably the only) framework tool, that allows us to do that easily.

Reply:
Thanks Ian!! I'll update it.